ENH Allow for appropriate dtype us in `preprocessing.PolynomialFeatures` for sparse matrices

Micky774 opened the pull request:
#### Reference Issues/PRs
Fixes #16803
Fixes #17554
Resolves #19676 (stalled)
Resolves #20524 (stalled)

#### What does this implement/fix? Explain your changes.
PR #20524: Calculates number of non-zero terms for each degree (row-wise) and creates dense arrays for `data/indices/indptr` to pass to Cython `_csr_polynomial_expansion`. Since the size is known a-priori, the appropriate dtype can be used during construction. The use of fused types in `_csr_polynomial_expansion` allows for only the minimally sufficient index dtype to be used, decreasing wasted memory when `int32` is sufficient.

This PR: reconciles w/ main and makes minor changes.

#### Any other comments?
The full functionality of this PR is really only enabled in `scipy_version>1.8` since it depends on an upstream bug fix

niuk-a committed:
[WIP] FIX index overflow error in sparse matrix polynomial expansion (bis)

Fixes #19676

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Reconciled with main

Micky774 committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into csr_polynomial

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Removed extra `total_nnz` assignment

Micky774 committed:
Added fused type

Micky774 committed:
Added clarifying comment

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Added changelog entry

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Fixed PR tag in changelog entry

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -890,6 +890,32 @@ def test_polynomial_features_deprecated_n_input_features():
+        m_index, np.array([119999, 120000, 7200180000, 119999, 120000, 7200180000])
+    )
+
+
```
Could you please write a test with `X.shape[1] == 65535` and `X.shape[1] == 65536` to check the dtype switch with `interaction_only=False`?

```
X = csr_matrix(([1.0], [65535 - 1], [0, 1]))
```


The limit comes from:

```python
>>> 1/2 * (np.sqrt(1 + 8 * np.iinfo(np.int32).max) - 1)
65535.49998664856
```

(I cheated with https://www.wolframalpha.com/input?i=%28d%5E2+%2B+d%29+%2F+2+%3D+e)

It could be relevant to do similar tests for `interaction_only=True` and `degree=3`.

ogrisel reviewed (commented):
Thanks for the PR. Here is a first pass of feedback.

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -394,17 +401,55 @@ def transform(self, X):
                 XP = sparse.csr_matrix((n_samples, 0), dtype=X.dtype)
             else:
-                XP = sparse.hstack(to_stack, format="csr")
+                XP = sparse.hstack(to_stack, format="csr", dtype=X.dtype)
```
While working on the tests suggested by @ogrisel , I came upon an unfortunate problem:

This line is problematic, since it turns out that during the `hstack` call, all matrices in `to_stack` get first converted to `COO` matrices, which automatically creates a `col` `ndarray` of `dtype=np.int32` if small enough. 

For example, if `n_features=65535, interaction_only=False, include_bias=False, degree=2` then there are two matrices w/ `65535` and `2147450880` columns respectively. Both are smaller than `np.iinfo(np.int32).max` so the resulting `COO` matrices have `col` arrays of type `np.int32`.

The problem is then when the offset is calculated for concatenation, e.g. `2147450880+65535` is calculated, which results in an integer overflow for `np.int32` leading to negative column indices.

This looks like a solid upstream issue, and I'm not sure what the best way around this would be. Perhaps implementing our own narrow-scope `hstack`?

ogrisel replied:
Maybe we can skip those new tests when `sp_version < parse_version("1.8")` then.

Micky774 committed:
Streamlined logic and improved tests

Micky774 committed:
Added test depending on scipy version

Micky774 committed:
Clarified breaking and renamed types

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Merge branch 'main' into csr_polynomial

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -384,6 +390,11 @@ def transform(self, X):
+            # matrix must be, and can then use this in construction the pre-
+            # stack matrices.
+            max_int32 = np.iinfo(np.int32).max
+            index_t = np.int64 if self.n_output_features_ > max_int32 else np.int32
```
Please rename `index_t` to `index_dtype` as this is a dtype object rather than a simple C type.

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -393,18 +404,56 @@ def transform(self, X):
+            # Covnert current stack to valid indices/indptr types
+            for mat in to_stack:
+                mat.indices = mat.indices.astype(index_t, copy=False)
+                mat.indptr = mat.indptr.astype(index_t, copy=False)
```
Is this really needed? `to_stack` only has the optional intercept column and `X` at this point.

Also we should not mutate the input matrix `X` silently. This can cause very weird to debug issues in case scikit-learn is used a multi-threaded program for instance.

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -393,18 +404,56 @@ def transform(self, X):
+                if expanded_d == 0:
                     break
-                to_stack.append(Xp_next)
+                assert expanded_d > 0
```
As a safety, we should also check that if `total_nnz > max_int32` then we should use `index_t = np.int64` for the subsequent index arrays.

ogrisel replied:
This was originally considered in https://github.com/scikit-learn/scikit-learn/pull/23731#discussion_r907128501 which was marked as resolved without being fully resolved.

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -890,6 +891,49 @@ def test_polynomial_features_deprecated_n_input_features():
+    assert xinter.dtype == X.dtype
+    assert xinter.indptr.dtype == xinter.indices.dtype == np.int32
+
+    # Depends on upstream bug fix
```
Should we `xfail` in the other case?
```suggestion
    # Breaks when `n_output_features_ > max_int32` for scipy
    # versions earlier than 1.8.0
    # Depends on upstream bug fix
```

jjerphan reviewed (commented):
Thank you, @Micky774.

A few comments.

Note that a specific dtype and C type have been added for sparse matrices indices.

https://github.com/scikit-learn/scikit-learn/blob/b157ac718754a235e443f3f5671fda7d2cd1dbba/sklearn/utils/_typedefs.pxd#L19-L28

Could we propagate this here for semantics?

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 commented:
I'm in the process of addressing the remaining review feedback, but it currently looks like there's another edge-case that `sparse.hstack` doesn't account for: when the individual `indices` arrays contain values less than `max_int32` but their concatenation would push it over the threshold. Once I pin down what the problem is exactly I may open an issue/PR upstream for this.

Micky774 committed:
Improved tests

Micky774 committed:
Merge branch 'csr_polynomial' of https://github.com/Micky774/scikit-learn into csr_polynomial

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Initial addition -- fails with segfault

Micky774 committed:
Improved documentation

Micky774 committed:
Added license information

Micky774 commented:
I've introduced `_csr_hstack.pyx` into the mix. This essentially reimplements (a very specific/limited-scope version of) scipy's `sparse.hstack` function. The logic isn't _too_ complex, and this bypasses the need to wait for the upstream change. Later on, if/when this is fixed upstream we could probably keep this as an internal "fix" to support users with a version of scipy that doesn't include the fix?

Micky774 commented on sklearn/preprocessing/_csr_hstack.pyx:
```diff
@@ -0,0 +1,103 @@
+                # We iterate over the valid indices, updating the indices and
+                # data
+                for kdx in range(row_start, row_end):
+                    indices[row_sum + kdx] = indices_bound[jdx][kdx] + offset
```
This line is generating a segmentation fault. Currently in the process of debugging, but if anyone has any ideas please let me know :)

thomasjpfan replied:
Can you post a code snippet that segfaulted?

Micky774 committed:
Updated fused-type name to hopefully clarify purpose

thomasjpfan commented on sklearn/preprocessing/_csr_hstack.pyx:
```diff
@@ -0,0 +1,103 @@
+
+# Fused type for data
+ctypedef fused DATA_t:
+    cython.numeric
```
This includes complex numbers as well. For our use case, I think we only need support for `float{32,64}`, `int{32,64}`

Micky774 committed:
Used vectors and updated implementation

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx:
```diff
@@ -0,0 +1,404 @@
+
+from sklearn.utils.fixes import threadpool_limits
+
+cpdef enum WeightingStrategy:
```
I think this file was accidentally committed.

Micky774 replied:
Good catch!

Micky774 committed:
Removed accidentally-added file

Micky774 committed:
Simplified and cleaned up implementation

Micky774 committed:
Slightly better formatting and variable name

Micky774 committed:
Fixed dtype bug and added testing

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Updated test to verify nnz count and indices

Micky774 committed:
Improved dtype resolution and clarified with comments

Micky774 commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -394,17 +401,49 @@ def transform(self, X):
                     break
-                to_stack.append(Xp_next)
+                assert expanded_d > 0
+                index_t = np.int64 if expanded_d > np.iinfo(np.int32).max else np.int32
```
Sorry for the late adoption, but this should be good now :)

Micky774 commented:
> Note that a specific dtype and C type have been added for sparse matrices indices.

Not sure how to incorporate this while retaining functionality -- the cython here requires `{32,64}bit` fused-types and there's not any place that would still work if instead using a single-type `SPARSE_INDEX_TYPE_t`. Sorry if I'm getting a bit confused 😅

Micky774 committed:
Fixed inexact index error

Micky774 committed:
Updated formatting

Micky774 committed:
Cleaner diff and blame history

jjerphan commented:
> Not sure how to incorporate this while retaining functionality -- the cython here requires {32,64}bit fused-types and there's not any place that would still work if instead using a single-type SPARSE_INDEX_TYPE_t. Sorry if I'm getting a bit confused sweat_smile

Yes, in retrospective the definition of `SPARSE_INDEX_TYPE_t` is confusing if we want to handle sparse matrices' indices types. As type and dtype definitions for Cython implementations' need to be reworked, I think it is fine hard-coding `cnp.int{32,64}_t` directly to support both sparse matrices cases without using `SPARSE_INDEX_TYPE_t`.

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Fixed overflow bug in expanded index calculation

Micky774 commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -26,8 +39,13 @@ cdef inline INDEX_T _deg2_column(INDEX_T d, INDEX_T i, INDEX_T j,
 
-cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
-                                 INDEX_T interaction_only) nogil:
+cdef inline cnp.int64_t _deg3_column(
+    cnp.int64_t d,
+    cnp.int64_t i,
+    cnp.int64_t j,
+    cnp.int64_t k,
+    FLAG_t interaction_only
+    ) nogil:
```
I think the greatest pain point of the current implementation is that in order to compute the output index we must compute squares or cubes of indices (e.g. `i,j,k`) and while the output index may fit within `int{32,64}`, the intermediate square/cube calculation may overflow, throwing the entire thing off. This effectively limits the valid range of input indices.

I'm open to any suggestions for how to circumvent this issue.

jjerphan replied:
I think we can try using intermediary `int128`-typed variables for squared or cubes of indices.

NumPy does not defines `int128_t` but does define [`npy_int128`](https://github.com/numpy/numpy/blob/bb95cf015165bcf85f99ba0d28a7a6e727ca955b/numpy/__init__.pxd#L276) which should be available via `cnp.npy_int128` in Cython.

Micky774 replied:
Alright I ended up predicting when an intermediate overflow would occur, and then deferring to python in those cases (to make use of arbitrary-precision `PyLong`s). This _only_ occurs for the cases where the indices are too large to safely compute the output index, and so won't kick into effect unless the input data has billions of features in the best case, and 500k in the worst case. Even when it kicks in, the performance drop only affects exactly the elements with too-large indices, which I think is reasonable to assume would be the minority of the data.

<details>
<summary> Quick benchmarks: </summary>

```python
from sklearn.preprocessing import PolynomialFeatures
from scipy import sparse

X = sparse.random(100_000, 1000, random_state=0)
pf = PolynomialFeatures(interaction_only=False, include_bias=True, degree=2)
%timeit -n9 pf.fit_transform(X)
```
```
Cython only: 774 ms ± 7.6 ms per loop (mean ± std. dev. of 7 runs, 9 loops each)
PR:          784 ms ± 4.3 ms per loop (mean ± std. dev. of 7 runs, 9 loops each)
```
</details>

ogrisel replied:
Unfortunately the PR causes some serious GIL contention as demonstrated in the follow new version of the benchmark:

```python
from sklearn.preprocessing import PolynomialFeatures
from scipy import sparse
from concurrent.futures import ThreadPoolExecutor

tpe = ThreadPoolExecutor(max_workers=8)

X = sparse.random(100_000, 1000, random_state=0)
pf = PolynomialFeatures(interaction_only=False, include_bias=True, degree=2)
%timeit -n 3 list(tpe.map(pf.fit_transform, [X] * 2 * 8))
``` 

```
main: 793 ms ± 11 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
PR:   2.28 s ± 21.9 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
```

on a laptop with 8 physical cores.

I haven't checked the math but maybe there is a way to be more conservative w.r.t. using the Python arbitrary precision code?

ogrisel replied:
Or maybe we could go back to using `float64` arithmetic operations before a final integer cast to int64 as was done previously? Not sure if the rounding errors/lack of exact representability of all integers in the `float64` domain can be observed for this case.

AWNystrom replied:
> As NumPy does not defines `int128_t` and does define [`npy_int128` to be identical to `int64_t`](https://github.com/numpy/numpy/blob/bb95cf015165bcf85f99ba0d28a7a6e727ca955b/numpy/__init__.pxd#L276), I think we can to compute squared or cubes of indices try using `np.uint64_t`-typed variables (which are `unsigned long long int`-typed variables)
> 
> Alternatively, we could try to rework the formula not to have huge values.

I'm sure the formulas could be rewritten to be more computationally-friendly. I was going for simplicity.

Micky774 replied:
> Unfortunately the PR causes some serious GIL contention as demonstrated in the follow new version of the benchmark:

I'm not sure that the original problem was indeed a GIL contention, rather just a general slowdown due to the introduction of `_calc_expanded_dimensions` which was an expensive and repeated calculation done entirely in Python. I've since moved it into the Cython layer (and renamed to `_calc_expanded_nnz`). Now the times of your script are as follow:

```
main: 642 ms ± 25.3 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
PR: 644 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 3 loops each)
```

Micky774 committed:
Fix intermediate calculation overflow and refactor tests

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Fixed duplicated changelog entries

Micky774 commented:
@ogrisel @jjerphan This should be ready for review again if you'd be interested.

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -41,19 +83,38 @@ cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
-                              cnp.ndarray[INDEX_T, ndim=1] indptr,
-                              INDEX_T d, INDEX_T interaction_only,
-                              INDEX_T degree):
+def _csr_polynomial_expansion(
+    cnp.ndarray[DATA_t, ndim=1] data,           # TODO: Make const in Cython 3
+    cnp.ndarray[INDEX_A_t, ndim=1] indices,     # TODO: Make const in Cython 3
+    cnp.ndarray[INDEX_A_t, ndim=1] indptr,      # TODO: Make const in Cython 3
+    INDEX_A_t d,
+    cnp.ndarray[DATA_t, ndim=1] result_data,
+    cnp.ndarray[INDEX_B_t, ndim=1] result_indices,
+    cnp.ndarray[INDEX_B_t, ndim=1] result_indptr,
+    FLAG_t interaction_only,
+    FLAG_t degree
+):
```
For now, we can use fused-typed memoryview without the `const`-qualification.
```suggestion
# TODO: `const`-qualify data, indices and indptr when Cython>=3.0 is released.
def _csr_polynomial_expansion(
    DATA_t[:] data,                 # IN READ-ONLY
    INDEX_A_t[:] indices,           # IN READ-ONLY
    INDEX_A_t[:] indptr,            # IN READ-ONLY
    INDEX_A_t d,
    DATA_t[:] result_data,          # OUT
    INDEX_B_t[:] result_indices,    # OUT
    INDEX_B_t[:] result_indptr,     # OUT
    FLAG_t interaction_only,
    FLAG_t degree
):
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -25,9 +51,25 @@ cdef inline INDEX_T _deg2_column(INDEX_T d, INDEX_T i, INDEX_T j,
     else:
         return d * i - (i**2 + i) / 2 + j
 
+def py_deg2_column(d, i, j, interaction_only):
+    """Compute the index of the column for a degree 2 expansion
+
+    d is the dimensionality of the input data, i and j are the indices
+    for the columns involved in the expansion.
+    """
```
```suggestion
def py_deg2_column(d, i, j, interaction_only):
    """Compute the index of the column for a degree 2 expansion.

    d is the dimensionality of the input data, i and j are the indices
    for the columns involved in the expansion.

    This function is defined in Python to use `PyLong`'s arbitrary
    precision, and is called in Cython when C types' representability
    range is too small.
    """
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -41,19 +83,38 @@ cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
         return ((3 * d**2 * i - 3 * d * i**2 + i ** 3 - i
                  - 3 * j**2 - 3 * j) / 6
                 + d * j + k)
+def py_deg3_column(d, i, j, k, interaction_only):
+    """Compute the index of the column for a degree 3 expansion
 
+    d is the dimensionality of the input data, i, j and k are the indices
+    for the columns involved in the expansion.
+    """
```
```suggestion
def py_deg3_column(d, i, j, k, interaction_only):
    """Compute the index of the column for a degree 3 expansion.

    d is the dimensionality of the input data, i, j and k are the indices
    for the columns involved in the expansion.

    This function is defined in Python to use `PyLong`'s arbitrary
    precision, and is called in Cython when C types' representability
    range is too small.
    """
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -128,9 +167,13 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                        col = _deg2_column(d, i, j, interaction_only)
-                        expanded_indices[expanded_index] = col
-                        expanded_data[expanded_index] = (
+                        if max(i, j) > MAX_SAFE_INDEX_DEG2:
+                            with gil:
+                                col = <INDEX_B_t> py_deg2_column(d, i, j, interaction_only)
+                        else:
+                            col = <INDEX_B_t> _deg2_column(d, i, j, interaction_only)
```
```suggestion
                        if max(i, j) > MAX_SAFE_INDEX_DEG2:
                            # In this case, the Cython implementation
                            # would integer overflow. 
                            # Here, we use arbitrary precision.
                            with gil:
                                col = <INDEX_B_t> py_deg2_column(d, i, j, interaction_only)
                        else:
                            col = <INDEX_B_t> _deg2_column(d, i, j, interaction_only)
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -139,14 +182,15 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                            col = _deg3_column(d, i, j, k, interaction_only)
-                            expanded_indices[expanded_index] = col
-                            expanded_data[expanded_index] = (
+                            if max(i, j, k) > MAX_SAFE_INDEX_DEG3:
+                                with gil:
+                                    col = <INDEX_B_t> py_deg3_column(d, i, j, k, interaction_only)
+                            else:
+                                col = <INDEX_B_t> _deg3_column(d, i, j, k, interaction_only)
```
```suggestion
                            if max(i, j, k) > MAX_SAFE_INDEX_DEG3:
                                # In this case, the Cython implementation
                                # would integer overflow. 
                                # Here, we use arbitrary precision.
                                with gil:
                                    col = <INDEX_B_t> py_deg3_column(d, i, j, k, interaction_only)
                            else:
                                col = <INDEX_B_t> _deg3_column(d, i, j, k, interaction_only)
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,20 +1,46 @@
 # Author: Andrew nystrom <awnystrom@gmail.com>
 
```
I think you can add yourself as an author here (I am not suggesting because I do not know which email you would like to use).

jjerphan reviewed (commented):
Thank you, @Micky774.

Here is a first review. 

Potentially this PR can be split into two PR:
 - a first one for the fix for `hstack` 
 - a second one for the fix for `preprocessing.PolynomialFeatures`.

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
Update comment for scipy min version (new backport)

Micky774 committed:
Removed vendored csr_hstack and instead error where appropriate

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Updated error message

Micky774 committed:
CLN Add authorship and delete cosmetic changes

jjerphan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -841,13 +841,31 @@ def test_csr_polynomial_expansion_index_overflow_deg2(
+    # `scipy.sparse.hstack` breaks in scipy<1.9.2 when
+    # `n_output_features_ > max_int32`. Test that we provide
+    # an informative error message.
+    if (
+        sp_version < parse_version("1.9.2")
+        and n_features == 65535
+        and not interaction_only
+    ):
```
```suggestion
    if (
        sp_version < parse_version("1.9.2")
        and n_features >= 65535
        and not interaction_only
    ):
```

Micky774 replied:
Actually the condition is a bit more specific, since if `n_features` is sufficiently large then this error won't be raised (component matrices will have `idx_dtype=int64`). Since the case is specifically designed to be `65535` in the test I'd rather keep it as-is. I think it is fine, since the causes are documented in the error and this test value specifically guarantees it. What do you think?

jjerphan replied:
This works for me. I think we can include an inline comment based on your remark:

```suggestion
    if (
        sp_version < parse_version("1.9.2")
        # When `n_features>65535`, `scipy.sparse.hstack` use the appropriate dtype
        # for representing indices and indptr.
        # `n_features==65535` is a corner case that we need to test.
        and n_features == 65535
        and not interaction_only
    ):
```

What do you think?

jjerphan commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -497,12 +419,26 @@ def transform(self, X):
+                        " 32bit integers results in negative columns. To avoid this"
+                        " error, either use a version of scipy `>=1.9.2` or alter the"
+                        " `PolynomialFeatures` transformer to produce fewer output"
+                        " features."
```
A suggestion to mention that using `np.int64` for `indices` and `indptr` should work
```suggestion
                        "Due to a bug in `scipy.sparse.hstack` present in SciPy<1.9.2,"
                        " stacking sparse matrices such that the resulting matrix would"
                        " have its `n_cols` too large to be represented by 32bit
                        " integers results in negative column indices.\n"
                        " To avoid this error, either use `scipy>=1.9.2`, convert your"
                        " input matrix `indices` and `indptr` array to use `np.int64`
                        " instead of `np.int32`, or alter the `PolynomialFeatures`
                        " transformer to produce fewer output features."
```

Micky774 replied:
> A suggestion to mention that using np.int64 for indices and indptr should work

Actually, I'm not sure that would work due to the eager downcasting of  CSR matrices upon construction. If all the `indices, indptr` of the CSR matrices to be stacked are `int64` but _could be_ `int32` then the `hstack` function will downcast them, so it only works if, for example, they have a value large enough to _require_ `int64`.

jjerphan replied:
Ah yes, I missed the downcast in `_stack_along_minor_axis` which is called in `bmat`. You can ignore my comment.

jjerphan reviewed (approved):
LGTM when the CI is green.

Since the scope of the PR has changed, I think this can be renamed to, for instance:

```
FIX Err in `preprocessing.PolynomialFeatures` if the output's indices and indptr are not representable
```

Micky774 committed:
Update sklearn/preprocessing/tests/test_polynomial.py

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
Revert

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 commented:
@ogrisel Wondering if you could take another look now that we've restricted the scope.

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,20 +1,46 @@
-ctypedef cnp.int32_t INDEX_T
 
-ctypedef fused DATA_T:
+cdef extern from "limits.h":
+    cdef int INT_MAX
+    cdef long LONG_MAX
```
Based [the eponymous pxd file for Cython `0.29.x`](https://github.com/cython/cython/blob/0.29.x/Cython/Includes/libc/limits.pxd), those constants should be `cimport`-able from it:
```suggestion
from lib.limits cimport INT_MAX, LONG_MAX
```

ogrisel commented:
Maybe @AWNystrom would be interested in having a quick look at this PR?

ogrisel reviewed (commented):
I think we need to find a way to solve this concurrent performance regression before considering the merge of this PR:

https://github.com/scikit-learn/scikit-learn/pull/23731/files#r1003581377

AWNystrom commented:
> Maybe @AWNystrom would be interested in having a quick look at this PR (as the original author of this code)?

Taking a look...

AWNystrom commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -25,9 +51,29 @@ cdef inline INDEX_T _deg2_column(INDEX_T d, INDEX_T i, INDEX_T j,
+    range is too small.
+    """
+    if interaction_only:
+        return d * i - (i**2 + 3 * i) // 2 - 1 + j
```
These are guaranteed to be divisible by 2, so are the // needed?

Micky774 replied:
This ensures the output is an `int` not a `float`

AWNystrom commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -41,19 +87,43 @@ cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
+    FLAG_t degree
+):
     """
     Perform a second-degree polynomial or interaction expansion on a scipy
```
Pre-existing (from me), but this function can do either second or third degree expansions, not just second.

AWNystrom commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -128,9 +179,16 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                        expanded_data[expanded_index] = (
+                        if max(i, j) > MAX_SAFE_INDEX_DEG2:
+                            # In this case, the Cython implementation
+                            # would integer overflow.
```
Maybe "would result in integer overflow." ?

AWNystrom reviewed (commented):
Love the changes. Huge thanks!

Micky774 committed:
Moved calculation of number of non-zero elements to Cython

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Addressed misc review feedback

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 commented:
Thanks for the feedback @AWNystrom! Regarding your comment on the equations -- I don't know that they can be simplified in a useful way. The main problem is overflows in intermediate terms of the calculation, and I think no matter how we reform the equation, so long as we retain precision there will be the same bottleneck. I think it's fine though with the recent improvements to efficiency though.

@ogrisel Wondering if you could take a look again. The concurrent performance regression should be addressed ([here](https://github.com/scikit-learn/scikit-learn/pull/23731/files#r1005033361))

Micky774 committed:
Added format specification

Micky774 committed:
Added explicit equation used to generate constants

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -128,9 +219,16 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                        col = _deg2_column(d, i, j, interaction_only)
-                        expanded_indices[expanded_index] = col
-                        expanded_data[expanded_index] = (
+                        if max(i, j) > MAX_SAFE_INDEX_DEG2:
```
Does `d` need to be a part of this condition?

If `d = intmax32` and `i = 2`, then `d * i` will overflow and `i**2 + 3*i` will not.

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,20 +1,56 @@
+cdef cnp.int64_t MAX_SAFE_INDEX_DEG3 = <cnp.int64_t> pow(LONG_MAX, 1/3)/4
+
+# This is the maximum value before the intermediate computation
+# d**2 + d blows out
```
I think "overflow" is a better phrase here.

```suggestion
# d**2 + d overflows
```

This comment applies in other places.

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,20 +1,56 @@
+
+ctypedef cnp.int8_t FLAG_t
+
+# This is (approximately) the maximum value before the intermediate computation
+# i**2 + 3*i blows out.
+# (Approximate) solution to i**2 + 3*i = maxint32
+cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4
```
Can we say this is an conservative upper bound for overflow?

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Improved documentation and introduced error

Micky774 committed:
Improved wording

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Overhauled tests

Micky774 commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -343,29 +344,93 @@ def transform(self, X):
                 to_stack.append(X)
+
             for deg in range(max(2, self._min_degree), self._max_degree + 1):
-                Xp_next = _csr_polynomial_expansion(
```
The calculations for `total_nnz` and `extended_col` have been moved back to Cython, but having those available in python for early stopping and for `dtype` resolution are helpful. The creation of the containers also needs to happen one layer above of the actual `_csr_polynomial_expansion` for fused-type resolution. It's honestly much easier in the python layer right now, though we could perhaps refactor it into a separate private method if you'd prefer?

thomasjpfan replied:
What do you think of a new Python function in `sklearn/preprocessing/_polynomial.py` to handle the logic in the loop?

Micky774 committed:
Improved cython routines adressed feedback

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -81,72 +208,52 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                        expanded_indices[expanded_index] = col
-                        expanded_data[expanded_index] = (
-                            data[i_ptr] * data[j_ptr])
+        row_starts = indptr[row_i]
```
I think it's important to keep the `with nogil` here.

Micky774 replied:
Good catch

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -81,72 +208,52 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
+                        # In this case, the Cython implementation
+                        # would result in an integer overflow.
+                        # Here, we take advantage of `PyLong` for arbitrary precision.
+                        with gil:
+                            col = <INDEX_B_t> py_deg2_column(n_features, i, j, interaction_only)
+                    else:
+                        col = <INDEX_B_t> _deg2_column(n_features, i, j, interaction_only)
```
Is it possible to move the complexity into `_deg2_column`? In other words, `_deg2_column` will check the condition and go to `py_deg2_column` when it needs to.

My main reason for this is to have a code comment in `_deg2_column` that explains the heuristic for switching. I also prefer to see the constant defined closer to `_deg2_column` so it is easier to reason about.

```python
# explain condition
cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4

cdef inline void _deg2_column(..., INDEX_B_t* col_out) nogil:
    if max(i, y) > MAX_SAFE_INDEX_DEG2:
       with gil:
           col_out[0] = py_deg2_column(...)
    ...
```

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,59 +1,174 @@
 
-ctypedef fused DATA_T:
+# This is a conservative upper bound for the maximum value before the
+# intermediate computation 3 * d**2 * d + d**3 overflows. We leverage the fact
```
As discussed IRL, I think we need a comment that connects the equation in `_deg3_column` to this approximate bound.

ogrisel commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,59 +1,174 @@
+# This is a conservative upper bound for the maximum value before the
+# intermediate computation 3 * d**2 * d + d**3 overflows. We leverage the fact
+# that i<=d.
+# Corresponds to 3 * d**2 * d + d**3 = maxint32
```
```suggestion
# Corresponds to 3 * n_features**2 * d + d**3 = maxint32
# when n_features == d.
```

Micky774 committed:
Improved code organization

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Opted for explicit `cnp.*` typing for `DATA_t`

Micky774 committed:
Reverted extraneous change

Micky774 committed:
Adjusted tests for un-indexable values on 32bit systems

jjerphan commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -27,6 +32,53 @@
+        return
+    assert expanded_col > 0
+
+    # This only checks whether each block needs 64bit integers upon
+    # expansion. We prefer to keep 32bit integers where we can,
+    # since currently CSR construction downcasts when possible, so
+    # we'n_features prefer to avoid an unnecessary cast. The dtype may still
+    # change in the concatenation process if needed.
```
Are there typos here?
```suggestion
    # This only checks whether each block needs 64bit integers upon
    # expansion. We prefer to keep 32bit integers where we can,
    # since currently SciPy's CSR construction downcasts when possible, 
    # so we prefer to avoid an unnecessary cast. The dtype may still
    # change in the concatenation process if needed.
```

jjerphan reviewed (dismissed):
This PR still LGTM. I let the other reviewers re-review and merge when it looks good for them. Thank you, @Micky774!

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,190 @@
+
+    # This is a conservative upper bound for the maximum value before the
+    # intermediate computation i**2 + 3*i overflows.
+    # Corresponds to to i**2 + 3*i = maxint32
+    cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4
```
We can hardcode the value instead of computing it everytime:

```suggestion
    # Corresponds to to i**2 + 3*i = maxint32
    # i = sqrt(maxint32) - 4
    cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = 46340
```

This comments applies to the other bounds.

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,190 @@
+    # Corresponds to to i**2 + 3*i = maxint32
+    cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4
+
+    if max(i, j) > MAX_SAFE_INDEX_DEG2 or i > LONG_MAX // n_features :
```
A comment about the second condition relating to `LONG_MAX // n_features`?

```suggestion
    if max(i, j) > MAX_SAFE_INDEX_DEG2 or i > LONG_MAX // n_features:
```

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,190 @@
+    # when n_features == d.
+    cdef cnp.int64_t MAX_SAFE_INDEX_DEG3 = <cnp.int64_t> pow(LONG_MAX, 1/3)/4
+
+    if max(i, j, k) > MAX_SAFE_INDEX_DEG3 or i // 3 + 1 > LONG_MAX // n_features // n_features:
```
A comment about the second condition?

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Addressed Cython bug

Micky774 committed:
Added documentation for secondary checks

Micky774 committed:
Update sklearn/preprocessing/_csr_polynomial_expansion.pyx

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
Formatting

Micky774 committed:
Merge branch 'csr_polynomial' of https://github.com/Micky774/scikit-learn into csr_polynomial

thomasjpfan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,197 @@
+    # that i<=d.
+    # Corresponds to 3 * n_features**2 * d + d**3 = maxint32
+    # when n_features == d.
+    cdef cnp.int64_t MAX_SAFE_INDEX_DEG3 = <cnp.int64_t> pow(LONG_MAX, 1/3)/4
```
Is the `/4` in the correct place?

```suggestion
    cdef cnp.int64_t MAX_SAFE_INDEX_DEG3 = <cnp.int64_t> pow(LONG_MAX / 4, 1/3)
```

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Update sklearn/preprocessing/_csr_polynomial_expansion.pyx

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 commented:
Regarding the risk of overflow in the intermediate calculations, would it be worth vendoring a portion of the `boost` library for `c++`? They have a nice `128bit` integer implementation which would alleviate our problems, and it uses native typing when possible (e.g. on `64bit` systems) so in those cases there's little-to-no overhead:

https://www.boost.org/doc/libs/1_81_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html

The portion of the library needed (plus license) is ~2.3MB.

Micky774 committed:
Factored equations to mitigate overflow risks

jjerphan commented:
Hi Meekail,

Thank you for doing some research.

I think vendoring such a library has a significant maintenance cost and has some packaging cost.

The current solution implicating code duplication looks like a proper thread-off to me if we are sure that it accurately represents indices without overflow.

What do others think?

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Added `__int128` support when available

Micky774 committed:
Added back python computation for overflow protection

Micky774 committed:
Corrected for linux

Micky774 committed:
Added support for CLANG and improved documentation

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 commented:
As it stands in this PR, we do not support `int64` indices on `32-bit` systems, consequently limiting the maximum output size of expanded matrices. This induces an upper bound on `n_features` of input matrices, i.e. not so large that the result is unrepresentable with `intp_t`, cf.
https://github.com/scikit-learn/scikit-learn/blob/3e9238a639bdcd9c0c5e8180d030b5b593f2e1cb/sklearn/preprocessing/_polynomial.py#L350-L360

In practice, I believe that this greatly limits the possibility of overflow arising on `32-bit` systems when using `int64` for calculation (current PR behavior). We can similarly approach a "near-guarantee" on `64-bit` system by using `int128` integers for calculation whenever possible. I have implemented this using compiler intrinsics available to `gcc` and `Clang`, and otherwise defaulted to `int64`:
https://github.com/scikit-learn/scikit-learn/blob/3e9238a639bdcd9c0c5e8180d030b5b593f2e1cb/sklearn/preprocessing/_csr_polynomial_expansion.pyx#L19-L29

It's not perfect, but it is the best compromise between correctness and maintainability/complexity I can find right now. Open to thoughts and suggestions.

Micky774 committed:
Removed unreachable code

Micky774 committed:
Slight change in equation form

thomasjpfan reviewed (commented):
The CI errors look real and from overflow. Maybe the compilers used there do not have 128 integers? A quick way to check is to return `sizeof(ITYPE_t)` and test it:

```python
# _csr_polynomial_expansion.pyx
def _get_size_of_ITYPE_t():
    return sizeof(ITYPE_t)

# test_polynomial.py
def test_get_size_of_int128_t():
    assert _get_size_of_ITYPE_t() == 16
```

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Updated to include test to confirm expected integer width

Micky774 committed:
Updated fused types

Micky774 commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,167 @@
+ctypedef fused INDEX_A_t:
+    int
+    long
+    long long
```
Unfortunately, using `int{32, 64}_t` from `_typedefs.pxd` leads to the Cython bug, so instead I have updated the typing from
```
int
long
long long
```
to
```
signed int
signed long long
```
to stay consistent.

Micky774 committed:
Updated typing

Micky774 commented:
> The CI errors look real and from overflow. Maybe the compilers used there do not have 128 integers? A quick way to check is to return `sizeof(ITYPE_t)` and test it:
> 
> ```python
> # _csr_polynomial_expansion.pyx
> def _get_size_of_ITYPE_t():
>     return sizeof(ITYPE_t)
> 
> # test_polynomial.py
> def test_get_size_of_int128_t():
>     assert _get_size_of_ITYPE_t() == 16
> ```

We shouldn't _need_ them to be `int128` all the time, just _at least_ `int64`. Specifically, even when `sizeof(intp_t)<=8` and `sizeof(ITYPE_t)==8` this should still work (as tested on my local WSL2 instance hence why I am a bit confused by the test failures. I've confirmed it on `Linux 64-bit`, `Windows 64-bit` and the `debian_atlas` instance we test on.

Micky774 committed:
Included feedback, and caught/handled old scipy bug

Micky774 commented:
This PR should be good for another review. Major updates from recent reviews:
1. Renamed type in `.pyx` from `ITYPE_t` to `LARGEST_INT_t` to better reflect that it is not the usual semantics of `ITYPE_t` and is not a fixed-size type untile compile time.
2. Caught and handled another niche bug in `scipy.sparse.hstack` present in scipy versions `<1.8.0`
3. Simplified and imrpoved equations
4. Adjusted typing of index types to better reflect `_typedef` definitions (`int{32, 64}_t`)

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 commented:
@thomasjpfan @jjerphan @ogrisel 

Apparently instead of a slow-down, we get a _speedup_ in this PR :)

![581f0be0-cad8-4ec1-a2f5-8cb892089f34](https://user-images.githubusercontent.com/34613774/227662822-14cb9603-4716-4bbf-b7bc-02617d63798f.png)

Micky774 commented:
Also not sure what is going on with the code coverage for this right now -- it seems to be failing due to the tests not being tested?

jjerphan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +793,207 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
     assert_array_almost_equal(Xt_csr.A, Xt_dense)
 
 
+def test_csr_polynomial_expansion_index_overflow_non_regression():
```
I think this test must be at least parametrized on `interaction_only` and potentially on `include_bias`.

What do you think?

jjerphan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +793,207 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+    # When `n_features>=65535`, `scipy.sparse.hstack` may not use the right
+    # dtype for representing indices and indptr if `n_features` is still
+    # small enough so that each block matrix's indices and indptr arrays
+    # can be represnented with `np.int32`. We test `n_features==65535`
```
```suggestion
    # can be represented with `np.int32`. We test `n_features==65535`
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,166 @@
+# This should work for GCC and CLANG, but doesn't for MSVC. We prefer to use
+# 128-bit integers when possible because the intermediate calculations have a
+# non-trivial risk of overflow. It is, however, very unlikely to come up on an
+# average use case, hence 64-bit integers are "good enough".
```
```suggestion
# average use case, hence 64-bit integers (i.e. `long long`) are "good enough".
```

jjerphan commented on sklearn/preprocessing/_csr_polynomial_expansion.pyx:
```diff
@@ -1,73 +1,166 @@
+        typedef long long LARGEST_INT_t;
+    #endif
+    """
+    ctypedef long long LARGEST_INT_t
```
I think that we might want to add a test to check for the size of `LARGEST_INT_t`.

What do you think?

ogrisel replied:
I agree but I am not sure how to set the expectations.

Maybe we can write a testing support Cython helper such as:

```cython
cpdef LARGEST_INT_t _increment_large_integer(LARGEST_INT_t value):
    return value + 1
```

and then call this helper in a test such as:

```python
def test_largest_int():
    if sys.platform == "win32":
        # On Windows, scikit-learn is typically compiled with MSVC that
        # does not support int128 arithmetic (at the time of writing):
        # https://stackoverflow.com/a/6761962/163740
        # Therefore we expect `_increment_large_integer` to overflow in the
        # following case:
        assert (
            _increment_large_integer(np.iinfo(np.int64).max) == np.iinfo(np.int64).min
        )
    else:
        # An all other supported platforms, C/C++ compilers used to build
        # scikit-learn are expected to support int128 operations:
        incremented = _increment_large_integer(np.iinfo(np.int64).max)
        assert incremented != np.iinfo(np.int64).min
        assert incremented == np.iinfo(np.int64).max + 1

```

Micky774 replied:
Would it be sufficient just to check the `sizeof` the type is either `8` or `16` based on expectation?

ogrisel replied:
Indeed.

Micky774 replied:
I have added `test_sizeof_LARGEST_INT_t()` to address this

jjerphan commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -343,29 +421,53 @@ def transform(self, X):
                 to_stack.append(X)
+
+            cumulative_size = sum(mat.shape[1] for mat in to_stack)
+            # import pdb; pdb.set_trace()
```
```suggestion
```

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +793,207 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+
 def test_polynomial_features_behaviour_on_zero_degree():
     """Check that PolynomialFeatures raises error when degree=0 and include_bias=False,
     and output a single constant column when include_bias=True
```
Could you also please include a test that XFAIL on Windows (with MSVC) to document a case of unmitigated integer overflow in `PolynomialFeatures.fit_transform` that would work with `int64` indexed output as long as the C/C++ compilers allows us to use `int128` typed intermediate variables?

ogrisel replied:
Note that this is different from (a complement to) what I suggest in https://github.com/scikit-learn/scikit-learn/pull/23731#discussion_r1150287129: the former would check our assumptions on the behavior of the compilers we typically use to compile scikit-learn while the latter would check the impact of this limitation to check that it can actually impact the behavior of `PolynomialFeatures` on such platforms.

Micky774 replied:
I have added `test_csr_polynomial_expansion_windows_fail` to address this

ogrisel reviewed (commented):
Thanks for pushing this PR through. I think this is the good track. Just a few comments with suggested improvements for in-line comments and better testing of platform-specific edge-cases.

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +793,207 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        msg = (
+            "The output that would result from the current configuration is too large"
+            " to be indexed"
+        )
```
As codecov complains because our linux 32 bit CI config is not configured to collect and upload test coverage data:

https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=53305&view=logs&j=aabdcdc3-bb64-5414-b357-ed024fe8659e&t=b7b3ba55-d585-563b-a032-f235636c22b0

I think it would be great to change the CI config to do it, but maybe this is complex for some reason? Maybe @thomasjpfan remembers why we did not do it? I don't.

ogrisel replied:
Thanks @thomasjpfan.

It's weird that the codecov comments disappeared from the current PR... I swear I could see the previously and this is what prompted me to comment.

ogrisel replied:
I merged #26027 with more fixes to the codecov config. I also merged `main` into this PR to confirm this fixes the codecov reports and check all the branches are now covered.

ogrisel replied:
Ok coverage is really fixed in `main` now. I pushed another merge commit to this branch.

ogrisel replied:
This is now fixed.

Micky774 committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Micky774 committed:
Fixed typo

ogrisel committed:
Merge branch 'main' into csr_polynomial

ogrisel committed:
Merge branch 'main' into csr_polynomial

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -295,6 +362,17 @@ def fit(self, X, y=None):
+                f" currently [{self._min_degree}, {self._max_degree}]\n- Whether to"
+                f" include only interaction terms, currently {self.interaction_only}\n-"
+                f" Whether to include a bias term, currently {self.include_bias}."
+            )
```
This suggestion was not addressed. I think it's still valid.

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -343,29 +421,52 @@ def transform(self, X):
+                    and self.n_output_features_ > max_int32
+                    and all_int32
+                ):
+                    raise ValueError(
```
I think we can ignore the lack of test coverage for the `< 1.9.2` bug, unless you find a way to trigger this case without changing the CI config.

Micky774 replied:
@thomasjpfan and I discussed the possibility of monkeypatching the scipy version _down_ on CI w/ scipy `>1.9.2` to trigger the `ValueError`. Do you think it would be simpler to ignore the lack of coverage, or is it worth including the monkeypatch in what is already a fairly complex test?

ogrisel replied:
No opinion. Do as you wish.

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +793,214 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        if interaction_only:
+            return d * i - (i**2 + 3 * i) // 2 - 1 + j
+        else:
+            return d * i - (i**2 + i) // 2 + j
```
This branch of the other hand should be covered if the test is parametrized with `interaction_only` as @jjerphan suggested above.

Micky774 committed:
Update sklearn/preprocessing/_polynomial.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Improved tests

Micky774 committed:
Improved tests

Micky774 commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +793,207 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+    row_nonzero, col_nonzero = X_trans.nonzero()
+    second_degree_idx = (
+        degree_2_calc(n_features, n_features - 2, n_features - 1, True) + n_features
+    )
```
I have renamed the variables and reorganized calculations a bit to hopefully make the causality and logic clearer

Micky774 commented:
@ogrisel @thomasjpfan I have addressed the existing feedback and improved the tests as requested

Micky774 committed:
Clean paranthesis

thomasjpfan commented:
The CI failures look related to the changes in this PR.

Micky774 committed:
Updated test to account for 32 bit systems

Micky774 committed:
Updated ValueError match string

Micky774 committed:
Fixed overflow bug in tests for Windows

Micky774 commented:
@thomasjpfan it should be good now

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        return
+    X_trans = pf.fit_transform(X)
+    row_nonzero, col_nonzero = X_trans.nonzero()
+    n_degree_1_features = n_features
+    lower_degree_features = n_degree_1_features + include_bias
```
If I understand the semantics of "lower", can `lower_degree_features` be `n_degree_1_features?

```suggestion
    n_degree_1_features_out = n_features + include_bias
``

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+
+    assert pf.n_output_features_ == max_degree_2_idx + 1
+    assert X_trans.dtype == dtype
+    assert X_trans.shape == (13, max_degree_2_idx + 1)
```
```suggestion
    assert X_trans.shape == (n_samples, max_degree_2_idx + 1)
```

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        n_features - 1 + int(include_bias),
+    ]
+    # Second degree index
+    target_indices.append(int(n_features * (n_features + 1) // 2 + target_indices[0]))
```
All the input are ints:

```suggestion
    target_indices.append(n_features * (n_features + 1) // 2 + target_indices[0])
```

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+    # Second degree index
+    target_indices.append(int(n_features * (n_features + 1) // 2 + target_indices[0]))
+    # Third degree index
+    target_indices.append(
+        int(n_features * (n_features + 1) * (n_features + 2) // 6 + target_indices[1])
+    )
```
```suggestion
    target_indices.append(
        n_features * (n_features + 1) * (n_features + 2) // 6 + target_indices[1]
    )
```

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+        and n_features == 65535
+        and degree == 2
+        and not interaction_only
+    ):
```
For codecov:

```suggestion
    ):  # pragma: no cover
```

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+            pf.fit(X)
+        return
+
+    # On older versions of scipy, a bug occurs when an intermediate matrix in
```
```suggestion
    # In SciPy < 1.8, a bug occurs when an intermediate matrix in
```

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+    # On older versions of scipy, a bug occurs when an intermediate matrix in
+    # `to_stack` in `hstack` fits within int32 however would require int64 when
+    # combined with all previous matrices in `to_stack`.
+    has_bug = False
```
I think we only need to check for the bug when SciPy<=1.8? Concretely:

```python
    if sp_version < parse_version("1.8.0"):
        has_scipy_hstack_bug = False
        ...
        
        if has_scipy_hstack_bug:
            msg = r"..."
            with pytest.raises(...):
                pf.fit_transform(X)
```

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+    X_trans = pf.fit_transform(X)
+
+    expected_dtype = np.int64 if num_combinations > np.iinfo(np.int32).max else np.int32
+    higher_order_count = (degree - 1) * int(not interaction_only)
```
Does "higher order" mean "higher than degree one"?

thomasjpfan commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -788,6 +795,258 @@ def test_polynomial_features_csr_X_dim_edges(deg, dim, interaction_only):
+    assert X_trans.shape == (1, pf.n_output_features_)
+    assert X_trans.indptr.dtype == X_trans.indices.dtype == expected_dtype
+    assert X_trans.nnz == expected_nnz
+    for idx in range(higher_order_count):
+        assert X_trans[0, target_indices[idx]] == pytest.approx(1.0)
```
I do not think this checks for all the non-zero indices. Is there a way to quickly check all that the indices of all the non-zeros elements are correct?

Micky774 committed:
Adopted review feedback

Micky774 committed:
Improved constant documentation

Micky774 committed:
Improved variable names

thomasjpfan commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -295,6 +362,27 @@ def fit(self, X, y=None):
+            if (
+                np.intp == np.int32
+                and self.n_output_features_ <= np.iinfo(np.int64).max
+            ):
```
```suggestion
            ):  # pragma: nocover
```

thomasjpfan reviewed (approved):
LGTM

Micky774 committed:
Update sklearn/preprocessing/_polynomial.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Incorporated typedef changes

Micky774 commented:
@ogrisel wondering if you could take another look

ogrisel reviewed (commented):
Thanks for the update. Here are some more comments for a build problem that I am still investigating:

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -817,3 +1080,60 @@ def test_polynomial_features_behaviour_on_zero_degree():
+    # On Windows, scikit-learn is typically compiled with MSVC that
+    # does not support int128 arithmetic (at the time of writing):
+    # https://stackoverflow.com/a/6761962/163740
+    if sys.maxsize <= 2**32 or sys.platform == "win32":
```
Actually maybe the problem above was caused by previously generated .c files. I need to try again from a clean build folder to make sure.

ogrisel replied:
I tried again on 32 bit linux on armv7 because I could find more recent python / clang versions for that platform and now the build works. But the resulting `_csr_polynomial_expansion` compiled extension is still broken:

```python-traceback
root@4af20c1dad16:/io# pytest -lxk test_sizeof_LARGEST_INT_t sklearn/preprocessing/tests/test_polynomial.py 
ImportError while loading conftest '/io/sklearn/conftest.py'.
sklearn/conftest.py:18: in <module>
    from sklearn.datasets import fetch_20newsgroups
sklearn/datasets/__init__.py:8: in <module>
    from ._base import load_breast_cancer
sklearn/datasets/_base.py:20: in <module>
    from ..preprocessing import scale
sklearn/preprocessing/__init__.py:38: in <module>
    from ._polynomial import PolynomialFeatures
sklearn/preprocessing/_polynomial.py:22: in <module>
    from ._csr_polynomial_expansion import (
E   ImportError: /io/sklearn/preprocessing/_csr_polynomial_expansion.cpython-310-arm-linux-gnueabihf.so: undefined symbol: __divti3
```

Micky774 replied:
Indeed I think we need to first check for `32bit` systems before eagerly trying `__int128` or `_BitInt(128)`. Probably the most portable way would be to check `sizeof(void*) == 8` right? I'll implement that soon.

ogrisel replied:
Indeed:

> If you compile for a 32-bit architecture like ARM, or x86 with -m32, no 128-bit integer type is supported with even the newest version of any of these compilers. So you need to detect support before using, if it's possible for your code to work at all without it.

from: https://stackoverflow.com/a/54815033/163740

I think we should also link to that SO question which is quite a good reference for this topic.

ogrisel replied:
Please also add a TODO note such that once all platform have C23 compilers, we should instead switch to the new way to do this:

> SO C23 will let you typedef unsigned _BitInt(128)  u128, modeled on [clang's feature](https://clang.llvm.org/docs/LanguageExtensions.html#extended-integer-types) originally called _ExtInt() which works even on 32-bit machines; see [a brief intro to it](https://stackoverflow.com/questions/61411865/how-do-you-use-clangs-new-custom-size-int-feature). Current GCC -std=gnu2x doesn't even support that syntax yet.

This will require a few years, but then it should work both for 32-bit and 64-bit target architectures.

Note that the clang + 32-bit target is no just a theoretical possibility: emscrypten / WASM / pyodide is a concrete platform that is built on that combo. I think @lesteve has the required tooling to test this PR on that target.

lesteve replied:
Quick and dirty testing of this PR on Pyodide seems to indicate that the `test_sizeof_LARGEST_INT_t` fails but the other added tests pass:
```
>>> pytest.main(['--pyargs', 'sklearn.preprocessing', '-k', 'test_csr_polynomial_expansion_index_overflow_non_reg
ression or test_csr_polynomial_expansion_too_large_to_index or LARGEST_INT', '-v'])
============================= test session starts ==============================
platform emscripten -- Python 3.11.3, pytest-7.3.1, pluggy-1.0.0 -- /home/pyodide/this.program
cachedir: .pytest_cache
rootdir: /home/pyodide
collecting ... collected 1098 items / 1089 deselected / 9 selected                            
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[True-True] PASSED [ 11%]
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[True-False] PASSED [ 22%]
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[False-True] PASSED [ 33%]
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[False-False] PASSED [ 44%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[True-True] PASSED [ 55%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[True-False] PASSED [ 66%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[False-True] PASSED [ 77%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[False-False] PASSED [ 88%]
tests/test_polynomial.py::test_sizeof_LARGEST_INT_t FAILED               [100%]
=================================== FAILURES ===================================
__________________________ test_sizeof_LARGEST_INT_t ___________________________
    def test_sizeof_LARGEST_INT_t():
        # On Windows, scikit-learn is typically compiled with MSVC that
        # does not support int128 arithmetic (at the time of writing):
        # https://stackoverflow.com/a/6761962/163740
        if sys.maxsize <= 2**32 or sys.platform == "win32":
            expected_size = 8
        else:
            expected_size = 16
    
>       assert _get_sizeof_LARGEST_INT_t() == expected_size
E       assert 16 == 8
E        +  where 16 = _get_sizeof_LARGEST_INT_t()
/lib/python3.11/site-packages/sklearn/preprocessing/tests/test_polynomial.py:1094: AssertionError
=========================== short test summary info ============================
FAILED tests/test_polynomial.py::test_sizeof_LARGEST_INT_t - assert 16 == 8
================= 1 failed, 8 passed, 1089 deselected in 1.03s =================
```

> Note that the clang + 32-bit target is no just a theoretical possibility: emscripten / WASM / pyodide is a concrete platform that is built on that combo.

By the way I am not sure `__clang__` is defined in the Pyodide use case, I think `__EMSCRIPTEN__` is used instead.

Micky774 replied:
Okay so I have:
1. Added a check for `__EMSCRIPTEN__` as an alternative to `__clang__`
2. Updated the runtime test for largest integer type to account for the odd (but beneficial) exception that is emscripten

Micky774 replied:
Also when trying to build pydodide/emscripten for myself, I get the following error. I'm not sure what's going on here 🤷 :

<details>

```traceback

* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (Cython>=0.29.24, setuptools<=61.0, wheel)
* Getting dependencies for wheel...
Partial import of sklearn during the build process.
running egg_info
writing scikit_learn.egg-info/PKG-INFO
writing dependency_links to scikit_learn.egg-info/dependency_links.txt
writing requirements to scikit_learn.egg-info/requires.txt
writing top-level names to scikit_learn.egg-info/top_level.txt
reading manifest file 'scikit_learn.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '*' found under directory 'asv_benchmarks'
warning: no previously-included files matching '*' found under directory 'benchmarks'
warning: no previously-included files matching '*' found under directory 'build_tools'
warning: no previously-included files matching '*' found under directory 'maint_tools'
warning: no previously-included files matching '*' found under directory 'benchmarks'
warning: no previously-included files matching '*' found under directory '.binder'
warning: no previously-included files matching '*' found under directory '.circleci'
warning: no previously-included files found matching '.codecov.yml'
warning: no previously-included files found matching '.git-blame-ignore-revs'
warning: no previously-included files found matching '.mailmap'
warning: no previously-included files found matching '.pre-commit-config.yaml'
warning: no previously-included files found matching 'azure-pipelines.yml'
warning: no previously-included files found matching 'CODE_OF_CONDUCT.md'
warning: no previously-included files found matching 'CONTRIBUTING.md'
warning: no previously-included files found matching 'SECURITY.md'
warning: no previously-included files found matching 'PULL_REQUEST_TEMPLATE.md'
adding license file 'COPYING'
writing manifest file 'scikit_learn.egg-info/SOURCES.txt'
* Installing packages in isolated environment... (wheel)
* Building wheel...
Partial import of sklearn during the build process.
Traceback (most recent call last):
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/__init__.py", line 23, in <module>
    from . import multiarray
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 221, in check_package_status
    module = importlib.import_module(package)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/__init__.py", line 141, in <module>
    from . import core
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.11 from "/tmp/build-env-mnfj07nc/bin/python"
  * The NumPy version is: "1.24.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pep517/in_process/_in_process.py", line 351, in <module>
    main()
  File "/usr/local/lib/python3.11/site-packages/pep517/in_process/_in_process.py", line 333, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pep517/in_process/_in_process.py", line 249, in build_wheel
    return _build_backend().build_wheel(wheel_directory, config_settings,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 244, in build_wheel
    return self._build_with_temp_dir(['bdist_wheel'], '.whl',
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 229, in _build_with_temp_dir
    self.run_setup()
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 282, in run_setup
    self).run_setup(setup_script=setup_script)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 174, in run_setup
    exec(compile(code, __file__, 'exec'), locals())
  File "setup.py", line 655, in <module>
    setup_package()
  File "setup.py", line 645, in setup_package
    check_package_status("numpy", min_deps.NUMPY_MIN_VERSION)
  File "setup.py", line 248, in check_package_status
    raise ImportError(
ImportError: numpy is not installed.
scikit-learn requires numpy >= 1.17.3.
Installation instructions are available on the scikit-learn website: http://scikit-learn.org/stable/install.html


ERROR Backend subproccess exited when trying to invoke build_wheel
```
</details>

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Added check for 32bit-ness for clang

Micky774 committed:
Improved documentation

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Merge branch 'main' into csr_polynomial

Micky774 committed:
Updated for emscripten edge-case

lorentzenchr commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -27,6 +32,68 @@
+
+    if expanded_col == 0:
+        return None
+    assert expanded_col > 0
```
Can we please avoid `assert` statements in productive code, e.g. everywhere except tests?
Even some linters/code helpers complain about it, see https://beta.ruff.rs/docs/rules/assert/.

Micky774 replied:
Good catch. I went ahead and removed it entirely, rather than replacing with an exception, since we know that `_calc_expanded_nnz(n_features, ...)` is non-negative for `n_features>=0`

Micky774 committed:
Removed extraneous assertion

jjerphan commented:
@Micky774: is this PR reviewable now? If it is not the case, could you ping me when it is? :)

Micky774 commented:
> @Micky774: is this PR reviewable now? If it is not the case, could you ping me when it is? :)

It should be good for review! The last large point of interest was this thread: https://github.com/scikit-learn/scikit-learn/pull/23731#discussion_r1176190395

which _hopefully_ is resolved through: https://github.com/scikit-learn/scikit-learn/pull/23731#discussion_r1181661016

Let me know if you have any questions or concerns

ogrisel reviewed (approved):
Alright, LGTM. If the wasm env triggers a new failure we can fix it up in a follow-up PRs.

At some point we will introduce a pyodide CI entry but later :)

ogrisel commented:
Thanks @Micky774!

jjerphan commented:
Thank you all (and especially @Micky774) for your efforts in this rocambolesque journey!

Micky774 commented:
It was definitely quite a tale of bugs (old and new) and niche features/interactions 😅

I really appreciate the patience and support of the reviewers as well ❤️

ogrisel commented:
I tried to build scikit-learn in a local emscrypten sdk on macos and I get a build failure (seemingly unrelated to this PR, it fails when building `sklearn/utils/_typedefs.cpython-311-wasm32-emscripten.so`).

I suspect we might still want to test because, in retrospect the emscrypten condition seems a bit counterintuitive to me.
